examples/BERT/mlm_task.py

```diff
@@ -0,0 +1,263 @@
+        n = torch.cuda.device_count() // args.world_size
+        device = list(range(rank * n, (rank + 1) * n))
+    else:
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    import torchtext
+    if args.dataset == 'WikiText103':
+        from torchtext.experimental.datasets import WikiText103 as WLMDataset
+    elif args.dataset == 'WikiText2':
+        from torchtext.experimental.datasets import WikiText2 as WLMDataset
+    elif args.dataset == 'WMTNewsCrawl':
+        from data import WMTNewsCrawl as WLMDataset
+    elif args.dataset == 'EnWik9':
+        from torchtext.datasets import EnWik9
+    elif args.dataset == 'BookCorpus':
+        from data import BookCorpus
+    else:
+        print("dataset for MLM task is not supported")
+
+    try:
+        vocab = torch.load(args.save_vocab)
+    except:
+        train_dataset, test_dataset, valid_dataset = WLMDataset()
+        old_vocab = train_dataset.vocab
+        vocab = torchtext.vocab.Vocab(counter=old_vocab.freqs,
+                                      specials=['<unk>', '<pad>', '<MASK>'])
+        with open(args.save_vocab, 'wb') as f:
+            torch.save(vocab, f)
+
+    if args.dataset == 'WikiText103' or args.dataset == 'WikiText2':
+        train_dataset, test_dataset, valid_dataset = WLMDataset(vocab=vocab)
+    elif args.dataset == 'WMTNewsCrawl':
+        from torchtext.experimental.datasets import WikiText2
+        test_dataset, valid_dataset = WikiText2(vocab=vocab, data_select=('test', 'valid'))
+        train_dataset, = WLMDataset(vocab=vocab, data_select='train')
+    elif args.dataset == 'EnWik9':
+        enwik9 = EnWik9()
+        idx1, idx2 = int(len(enwik9) * 0.8), int(len(enwik9) * 0.9)
+        train_data = torch.tensor([vocab.stoi[_id]
+                                  for _id in enwik9[0:idx1]]).long()
+        val_data = torch.tensor([vocab.stoi[_id]
+                                 for _id in enwik9[idx1:idx2]]).long()
+        test_data = torch.tensor([vocab.stoi[_id]
+                                 for _id in enwik9[idx2:]]).long()
+        from torchtext.experimental.datasets import LanguageModelingDataset
+        train_dataset = LanguageModelingDataset(train_data, vocab)
+        valid_dataset = LanguageModelingDataset(val_data, vocab)
+        test_dataset = LanguageModelingDataset(test_data, vocab)
+    elif args.dataset == 'BookCorpus':
+        train_dataset, test_dataset, valid_dataset = BookCorpus(vocab)
+
+    train_data = process_raw_data(train_dataset.data, args)
+    if rank is not None:
+        # Chunk training data by rank for different gpus
+        chunk_len = len(train_data) // args.world_size
+        train_data = train_data[(rank * chunk_len):((rank + 1) * chunk_len)]
+    val_data = process_raw_data(valid_dataset.data, args)
+    test_data = process_raw_data(test_dataset.data, args)
+
+    if args.checkpoint != 'None':
+        model.bert_model = torch.load(args.checkpoint)
```

Comment:
Given that SLURM was used here, I would have also have expected to capture the sigterm/sigusr1 signal, as done [here](https://github.com/fairinternal/FAIR_Cluster/blob/master/checkpoint/main.py#L146). This allows creating a checkpoint when some signal is received, say when the job is about to timeout.